src/hb-kbts.cc

```diff
@@ -0,0 +1,248 @@
+  kbts_language kb_language = KBTS_LANGUAGE_DEFAULT;
+  {
+    hb_tag_t script;
+    hb_tag_t language;
+    unsigned int script_count = 1;
+    unsigned int language_count = 1;
+    hb_ot_tags_from_script_and_language (buffer->props.script, buffer->props.language,
+					 &script_count, &script,
+					 &language_count, &language);
+    if (script_count)
+      kb_script = kbts_ScriptTagToScript (hb_uint32_swap (script));
+    if (language_count)
+      kb_language = (kbts_language) hb_uint32_swap (language);
+  }
+
+  hb_vector_t<kbts_glyph> kb_glyphs;
+  if (unlikely (!kb_glyphs.resize_exact (buffer->len)))
+    return false;
+
+  for (size_t i = 0; i < buffer->len; ++i)
+    kb_glyphs.arrayZ[i] = kbts_CodepointToGlyph (kb_font, buffer->info[i].codepoint);
+
+  if (num_features)
+  {
+    for (unsigned int i = 0; i < num_features; ++i)
+    {
+      hb_feature_t feature = features[i];
+      for (unsigned int j = 0; j < kb_glyphs.length; ++j)
+      {
+	kbts_glyph *kb_glyph = &kb_glyphs.arrayZ[j];
+	if (hb_in_range (j, feature.start, feature.end))
+	{
+	  if (!kb_glyph->Config)
+	    kb_glyph->Config = (kbts_glyph_config *) hb_calloc (1, sizeof(kbts_glyph_config));
+	  kbts_glyph_config *config = kb_glyph->Config;
+	  while (!kbts_GlyphConfigOverrideFeatureFromTag (config, hb_uint32_swap (feature.tag),
+	                                                  feature.value > 1, feature.value))
+	  {
+	    config->FeatureOverrides = (kbts_feature_override *) hb_realloc (config->FeatureOverrides,
+									     config->RequiredFeatureOverrideCapacity);
+	    config->FeatureOverrideCapacity += 1;
+	  }
+	}
+      }
+    }
+  }
+
+  kbts_shape_state *kb_shape_state;
+  {
+    size_t kb_shape_state_size = kbts_SizeOfShapeState (kb_font);
+    void *kb_shape_state_buffer = hb_malloc (kb_shape_state_size);
+    kb_shape_state = kbts_PlaceShapeState (kb_shape_state_buffer, kb_shape_state_size);
+  }
+  kbts_shape_config kb_shape_config = kbts_ShapeConfig (kb_font, kb_script, kb_language);
+  uint32_t glyph_count = kb_glyphs.length;
+  uint32_t glyph_capacity = glyph_count;
+  while (kbts_Shape (kb_shape_state, &kb_shape_config, KBTS_DIRECTION_LTR, kb_direction,
+		     kb_glyphs.arrayZ, &glyph_count, glyph_capacity))
+  {
+    glyph_capacity = kb_shape_state->RequiredGlyphCapacity;
+    if (unlikely (!kb_glyphs.resize_exact (glyph_capacity)))
+      return false;
+  }
+
+  kb_glyphs.resize_exact (glyph_count);
+
+  hb_buffer_set_length (buffer, kb_glyphs.length);
+  hb_glyph_info_t *info = buffer->info;
+  hb_glyph_position_t *pos = buffer->pos;
+
+  float upem = hb_face_get_upem (face);
+  float x_mult = font->x_scale / upem;
+  float y_mult = font->y_scale / upem;
+
+  buffer->clear_positions ();
+  for (size_t i = 0; i < kb_glyphs.length; ++i)
+  {
+    kbts_glyph kb_glyph = kb_glyphs.arrayZ[i];
+    info[i].codepoint = kb_glyph.Id;
+    info[i].cluster = 0; // FIXME
+    pos[i].x_advance = roundf (kb_glyph.AdvanceX * x_mult);
```

Comment:
`font->em_scale_x`? Although, that function probably should be rewritten for int32 as well.

Reply:
`font->em_scalef_x` then?